src/trusted/service_runtime/linux/nacl_signal.c

```diff
@@ -179,7 +179,23 @@ static void FindAndRunHandler(int sig, siginfo_t *info, void *uc) {
        * the default behavior which is to exit the app with the signal
        * number as the error code.
        */
-      NaClExit(-sig);
+
+      /*
+      * Unset the TLS variable so that if a crash occurs during thread
+      * teardown, the signal handler does not dereference a dangling
+      * NaClAppThread pointer.
+      */
+      NaClTlsSetCurrentThread(NULL);
+      // We've logged its a trusted fault, lets cleanup
+      if (natp->is_cage_parent) {
+        if (!natp->tearing_down) AddToFaultTeardown(natp);
+      } else {
+        if (!natp->cage_parent->tearing_down) AddToFaultTeardown(natp->cage_parent);
+      }
+      
+      // just hang here while we cleanup
+      while (1);
+
     }
```

Comment:
I'm not a fan of busy loops. I'd much rather sleep or something like it.

Reply:
I actually can't call sleep while in the handler. Initially tried that.

Reply:
Can you do any other blocking call?